amads/core/vectors_sets.py

```diff
@@ -0,0 +1,294 @@
+
+    """
+    return [
+        x * scale_factor for x in input_list
+    ]  # TODO np would be better in cases like this
+
+
+# Transformations
+
+
+def rotate(vector: list[int] | tuple[int], steps: int | None = None) -> list:
+    """
+    Rotate a list by N steps.
+    This serves equivalently for
+    "phase shifting" of rhythm and
+    "transposition" of pitch.
+
+    Parameters
+    ----------
+    vector: Any list of any elements. We expect to work with a list integers representing a vector.
+    steps: how many steps to rotate.
+        Or, equivalently, the nth index of the input list becomes the 0th index of the new.
+        If unspecified, use the half cycle: int(<cycle lenth>/2).
+
+    Returns
+    -------
+    list | tuple: The input (list or tuple), rotated. Same length.
+
+    Examples
+    --------
+
+    >>> start = [0, 1, 2, 3]
+    >>> rotate(start, 1)
+    [1, 2, 3, 0]
+
+    >>> rotate(start, -1)
+    [3, 0, 1, 2]
+
+    >>> rotate([0, 1, 2, 3])  # note no steps specified
+    [2, 3, 0, 1]
+
+    """
+    if not steps:
+        steps = int(len(vector) / 2)
+
+    return vector[steps:] + vector[:steps]
+
+
+def mirror(vector: list | tuple, index_of_symmetry: int | None = None) -> list:
+    """
+    Reverse a list.
+
+    Parameters
+    ----------
+    vector: Any list of any elements. We expect to work with a list integers representing a vector.
+    index_of_symmetry: Defaults to None, in which case, standard refelction [::-1].
+        Alternatively, specify an index to rotate about, e.g., for the reverse function in convolution use 0.
+        This is equivalent to mirror and rotation.
+        See notes at `rotate`.
+
+    Returns
+    -------
+    list | tuple: The input (list or tuple), mirrored. Same length.
+
+    Examples
+    --------
+    >>> test_case = [0, 1, 2, 3, 4, 5]
+    >>> mirror(test_case)
+    [5, 4, 3, 2, 1, 0]
+
+    >>> mirror(test_case, index_of_symmetry=0)
+    [0, 5, 4, 3, 2, 1]
+
+    >>> mirror(test_case, index_of_symmetry=1)
+    [1, 0, 5, 4, 3, 2]
+    """
+    if index_of_symmetry is not None:
+        return vector[index_of_symmetry::-1] + vector[-1:index_of_symmetry:-1]
+    else:
+        return vector[::-1]
+
+
+def is_indicator_vector(indicator_vector: list | tuple) -> bool:
+    """
+    Check that a list or tuple is an indicator vector, featuring only 0s and 1s.
+    """
+    if all(x in (0, 1) for x in indicator_vector):
+        return True
+    return False
+
+
+def weighted_to_indicator(weighted_vector: list | tuple, threshold=0.0) -> tuple:
+    """
+    Converts a weighted vector to an indicator vector.
+
+    Parameters
+    ----------
+    weighted_vector: A NumPy array representing the weighted vector. TODO consider use of numpy
+    threshold:  Values below this threshold will be set to 0.  This handles
+    cases where weights might be very small but not exactly zero.
+
+    Returns
+    -------
+    A NumPy array representing the indicator vector (0s and 1s).
```

Comment:
The types specified in the docstring are inconsistent with the type hints in the function definition